fpsim/wrappers.py

```diff
@@ -0,0 +1,1216 @@
+2. **Duration changes**: Model how long people stay on a method (e.g., better counseling, 
+   reduced side effects, improved supply chain)
+3. **Method mix**: Model shifts in which methods people choose (e.g., social marketing campaigns)
+4. **Probability of use**: Model changes in overall contraceptive uptake
+5. **Switching patterns**: Model how people transition between methods
+
+Key concepts :
+---------------------------
+- **Efficacy**: How well the method prevents pregnancy (0 to 1, where 0.99 = 99% effective)
+- **Duration**: How many months people typically stay on the method before discontinuing
+- **Method mix**: The distribution of users across different contraceptive methods
+- **Switching matrix**: The probabilities of moving from one method to another
+
+Note: Small changes in duration can have large effects on method mix because people
+accumulate on methods they stay on longer. This is realistic - improving continuation
+rates is one of the most impactful interventions.
+
+Quick Reference 
+------------------------------
+**What intervention should I model for...**
+
+1. **Better counseling / side effect management** → `set_duration_months()`
+   People stay on methods longer with better support
+   
+2. **Improved product quality** → `set_efficacy()`
+   Better manufacturing or user education increases effectiveness
+   
+3. **Social marketing campaign for specific method** → `set_method_mix()`
+   Shift demand toward targeted methods (requires baseline capture)
+   
+4. **General awareness campaign** → `set_probability_of_use()`
+   Increase overall contraceptive use (CPR)
+   
+5. **Provider training on method transitions** → `scale_switching_matrix()`
+   Make it easier for people to switch to better methods
+   
+6. **Comprehensive program** → Combine multiple methods!
+   Example: Better counseling (duration) + quality improvements (efficacy)
+
+**Method name shortcuts:**
+- 'pill' = Pills
+- 'inj' = Injectables (DMPA, Sayana Press, etc.)
+- 'impl' = Implants (Jadelle, Implanon, etc.)
+- 'iud' = IUDs
+- 'cond' = Male condoms
+- 'btl' = Female sterilization (tubal ligation)
+- 'wdraw' = Withdrawal
+- 'othtrad' = Other traditional methods
+- 'othmod' = Other modern methods
+"""
+# endregion: Introduction
+
+import json
+from contextlib import contextmanager
+from dataclasses import dataclass, field
+from enum import Enum
+from functools import wraps
+from operator import attrgetter
+from pathlib import Path
+from typing import Dict, Optional, Sequence, Union, TypedDict, List, Any
+
+import numpy as np
+import sciris as sc
+
+from . import interventions as fpi
+from .methods import NewMethodConfig
+
+__all__ = [
+    'MethodIntervention',
+    'make_update_methods',
+    'MethodName',
+]
+
+# ==========================================
+# Enums and Type Definitions
+# ==========================================
+class MethodName(str, Enum):
```

Comment:
Should probabaly avoid defining these in >1 place. They're currently defined in `make_method_list()` within methods.py.